File: autozig-ecs/src/zig/event.zig
const std = @import("std");

// 事件队列 - 双缓冲实现
pub const EventQueue = struct {
    buffers: [2]std.ArrayList(u8),
    write_idx: u8,
    stride: usize, // 每个事件的字节大小
    allocator: std.mem.Allocator,
    
    pub fn init(allocator: std.mem.Allocator, event_size: usize) !*EventQueue {
        const queue = try allocator.create(EventQueue);
        queue.* = EventQueue{
            .buffers = [2]std.ArrayList(u8){
                std.ArrayList(u8){},
                std.ArrayList(u8){},
            },
            .write_idx = 0,
            .stride = event_size,
            .allocator = allocator,
        };
        return queue;
    }
    
    pub fn deinit(self: *EventQueue) void {
        self.buffers[0].deinit(self.allocator);
        self.buffers[1].deinit(self.allocator);
        self.allocator.destroy(self);
    }
    
    // 写入事件（写到当前写缓冲）
    pub fn push(self: *EventQueue, data: [*]const u8) !void {
        const buf = &self.buffers[self.write_idx];
        try buf.appendSlice(self.allocator, data[0..self.stride]);
    }
    
    // 交换缓冲区（每帧调用）
    pub fn swap(self: *EventQueue) void {
        self.write_idx ^= 1; // 0 <-> 1 切换
        self.buffers[self.write_idx].clearRetainingCapacity(); // 清空新的写缓冲
    }
    
    // 获取读缓冲（上一帧写入的）
    pub fn get_reader(self: *const EventQueue) []const u8 {
        const read_idx = self.write_idx ^ 1;
        return self.buffers[read_idx].items;
    }
    
    // 获取读缓冲中的事件数量
    pub fn get_event_count(self: *const EventQueue) usize {
        const read_idx = self.write_idx ^ 1;
        return self.buffers[read_idx].items.len / self.stride;
    }
    
    // 清空所有缓冲
    pub fn clear(self: *EventQueue) void {
        self.buffers[0].clearRetainingCapacity();
        self.buffers[1].clearRetainingCapacity();
    }
};

// 全局allocator在entity.zig中定义，合并后直接可用

// Exported C API
export fn event_queue_create(event_size: usize) ?*EventQueue {
    return EventQueue.init(g_allocator, event_size) catch null;
}

export fn event_queue_destroy(queue: *EventQueue) void {
    queue.deinit();
}

export fn event_queue_push(queue: *EventQueue, data_ptr: [*]const u8) bool {
    queue.push(data_ptr) catch return false;
    return true;
}

export fn event_queue_swap(queue: *EventQueue) void {
    queue.swap();
}

export fn event_queue_get_reader(queue: *const EventQueue, out_ptr: *[*]const u8, out_len: *usize) void {
    const reader = queue.get_reader();
    out_ptr.* = reader.ptr;
    out_len.* = reader.len;
}

export fn event_queue_get_event_count(queue: *const EventQueue) usize {
    return queue.get_event_count();
}

export fn event_queue_clear(queue: *EventQueue) void {
    queue.clear();
}

File: autozig-ecs/src/event.rs
//! Event system - 基于 Zig + Rust 的高性能事件系统

use autozig_macro::include_zig;
use crate::world::World;
use crate::resource::{Resource, ResMut};
use std::marker::PhantomData;

/// Event trait - 标记为事件的 trait
pub trait Event: Resource {}
impl<T: Resource> Event for T {}

#[repr(C)]
pub struct EventQueueOpaque {
    _private: u8,
}

include_zig!("src/zig/event.zig", {
    fn event_queue_create(event_size: usize) -> *mut EventQueueOpaque;
    fn event_queue_destroy(queue: *mut EventQueueOpaque);
    fn event_queue_push(queue: *mut EventQueueOpaque, event: *const std::ffi::c_void) -> bool;
    fn event_queue_clear(queue: *mut EventQueueOpaque);
    fn event_queue_swap(queue: *mut EventQueueOpaque);
    fn event_queue_get_reader(queue: *const EventQueueOpaque, out_ptr: *mut *const u8, out_len: *mut usize);
});

/// Bevy-compatible Events<E> resource
pub struct Events<E: Event> {
    pub(crate) queue: *mut EventQueueOpaque,
    _marker: PhantomData<E>,
}

impl<E: Event> Events<E> {
    pub fn new() -> Self {
        Self {
            queue: event_queue_create(std::mem::size_of::<E>()),
            _marker: PhantomData,
        }
    }

    pub fn send(&mut self, event: E) {
        let ptr = &event as *const E as *const std::ffi::c_void;
        event_queue_push(self.queue, ptr);
    }

    pub fn update(&mut self) {
        event_queue_swap(self.queue);
    }

    pub fn clear(&mut self) {
        event_queue_clear(self.queue);
    }
    
    pub fn get_reader(&self) -> EventReader<'_, E> {
        EventReader::new(self.queue)
    }

    pub fn get_writer(&mut self) -> EventWriter<'_, E> {
        EventWriter::new(self.queue)
    }

    pub fn is_empty(&self) -> bool {
        let mut ptr: *const u8 = std::ptr::null();
        let mut len: usize = 0;
        event_queue_get_reader(self.queue, &mut ptr, &mut len);
        len == 0
    }
}

impl<E: Event> Drop for Events<E> {
    fn drop(&mut self) {
        event_queue_destroy(self.queue);
    }
}

impl<E: Event> Default for Events<E> {
    fn default() -> Self {
        Self::new()
    }
}

unsafe impl<E: Event> Send for Events<E> {}
unsafe impl<E: Event> Sync for Events<E> {}

/// Bevy-compatible EventWriter
pub struct EventWriter<'w, E: Event> {
    queue: *mut EventQueueOpaque,
    _marker: PhantomData<&'w mut E>,
}

impl<'w, E: Event> EventWriter<'w, E> {
    pub(crate) fn new(queue: *mut EventQueueOpaque) -> Self {
        Self {
            queue,
            _marker: PhantomData,
        }
    }

    pub fn send(&mut self, event: E) {
        let ptr = &event as *const E as *const std::ffi::c_void;
        event_queue_push(self.queue, ptr);
    }
}

unsafe impl<'w, E: Event> Send for EventWriter<'w, E> {}
unsafe impl<'w, E: Event> Sync for EventWriter<'w, E> {}

/// Bevy-compatible EventReader
pub struct EventReader<'w, E: Event> {
    queue: *mut EventQueueOpaque,
    _marker: PhantomData<&'w E>,
}

impl<'w, E: Event> EventReader<'w, E> {
    pub(crate) fn new(queue: *mut EventQueueOpaque) -> Self {
        Self {
            queue,
            _marker: PhantomData,
        }
    }
    
    pub fn iter(&self) -> EventIter<'_, E> {
        let mut ptr: *const u8 = std::ptr::null();
        let mut len: usize = 0;
        event_queue_get_reader(self.queue, &mut ptr, &mut len);
        
        EventIter {
            data: if ptr.is_null() { &[] } else { 
                unsafe { std::slice::from_raw_parts(ptr as *const E, len / std::mem::size_of::<E>()) }
            },
            index: 0,
        }
    }
}

unsafe impl<'w, E: Event> Send for EventReader<'w, E> {}
unsafe impl<'w, E: Event> Sync for EventReader<'w, E> {}

pub struct EventIter<'a, E: Event> {
    data: &'a [E],
    index: usize,
}

impl<'a, E: Event> Iterator for EventIter<'a, E> {
    type Item = &'a E;
    fn next(&mut self) -> Option<Self::Item> {
        if self.index < self.data.len() {
            let item = &self.data[self.index];
            self.index += 1;
            Some(item)
        } else {
            None
        }
    }
}

pub struct EventId<E: Event>(u64, PhantomData<E>);
pub struct EventCursor<E: Event>(u64, PhantomData<E>);
pub struct EventRegistry;
pub struct EventParIter;

impl<E: Event> Resource for Events<E> {}

/// Event sent when the application should exit
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum AppExit {
    #[default]
    Success,
    Error(u8),
}

impl Resource for AppExit {}

/// Standard system to update events every frame
pub fn event_update_system<E: Event>(mut events: ResMut<Events<E>>) {
    events.update();
}


